src/L2Penalty_alg.jl

```diff
@@ -0,0 +1,431 @@
+export L2Penalty, L2PenaltySolver, solve!
+
+import SolverCore.solve!
+
+mutable struct L2PenaltySolver{T <: Real, V <: AbstractVector{T}, S <: AbstractOptimizationSolver} <: AbstractOptimizationSolver
+  x::V
+	s::V
+	s0::V
+  ψ::ShiftedCompositeNormL2
+	sub_ψ::CompositeNormL2
+	sub_solver::S
+	sub_stats::GenericExecutionStats{T, V, V, Any}
+end
+
+function L2PenaltySolver(
+  nlp::AbstractNLPModel{T, V};
+	sub_solver = R2Solver
+	) where{T, V}
+	x0 = nlp.meta.x0
+	x = similar(x0)
+	s = similar(x0)
+	s0 = zero(x0)
+
+	# Allocating variables for the ShiftedProximalOperator structure
+	(rows, cols) = jac_structure(nlp)
+  vals = similar(rows,eltype(x0))
+	A = SparseMatrixCOO(nlp.meta.ncon, nlp.meta.nvar, rows, cols, vals)
+	b = similar(x0, eltype(x0), nlp.meta.ncon)
+
+	
+	# Allocate ψ = ||c(x) + J(x)s|| to compute θ
+	ψ = ShiftedCompositeNormL2(1.0, 
+		(c, x) -> cons!(nlp, x, c), 
+		(j, x) -> jac_coord!(nlp, x, j.vals), 
+		A, 
+		b
+	)
+
+	# Allocate sub_ψ = ||c(x)|| to solve min f(x) + τ||c(x)||
+	sub_ψ = CompositeNormL2(1.0, 
```

Comment:
Why is this called `sub_ψ`?

Reply:
Because it is what the subsolver works with